### PR TITLE
adds wireguard stats

### PR DESCRIPTION
**Pull request by fernando-guerrero-leon**

### CODE REVIEW

File: encryptme_stats/metrics.py [hunk 1]
```diff
@@ -622,22 +650,100 @@ def _get_ipsec_session_stats():
...
+
+def wireguard():
+    """Gather Wireguard statistics."""
+    try:
+        interfaces = subprocess_out(["wg", "show", "interfaces"])
+        if not isinstance(interfaces, list):
+            return []
+        if interfaces and interfaces[-1] == '':
+            interfaces.pop()
+        info = []
```
Comment (dmibaranov) [inline encryptme_stats/metrics.py]: please declare at the top of function

File: encryptme_stats/metrics.py [hunk 2]
```diff
@@ -622,22 +650,100 @@ def _get_ipsec_session_stats():
...
+                'wireguard': {
+                    'interface': interface,
+                    'num_peers': num_peers,
+                    'public_key': public_key,
+                    'latest_handshake': latest_handshake,
+                }
+            })
+    except Exception as exc:
+        logging.debug("Error gathering wireguard stats: %s", exc)
+        return []
```
Comment (dmibaranov) [inline encryptme_stats/metrics.py]: please remove this line

File: encryptme_stats/metrics.py [hunk 3]
```diff
@@ -622,22 +650,100 @@ def _get_ipsec_session_stats():
...
+    return info
```
Comment (dmibaranov) [inline encryptme_stats/metrics.py]: missed newline

File: encryptme_stats/metrics.py [hunk 4]
```diff
@@ -80,23 +80,51 @@ def _get_openvpn_stats(path="/var/run/openvpn/server-0.sock"):
...
+    try:
+        now_epoch = int(datetime.utcnow().timestamp())
+        interfaces = subprocess_out(["wg", "show", "interfaces"])
+        if isinstance(interfaces, list):
+            if interfaces and interfaces[-1] == '':
+                interfaces.pop()
+            for interface in interfaces:
+                peers = subprocess_out(["wg", "show", interface, "dump"])
+                if isinstance(peers, list):
+                    if peers and peers[-1] == '':
```
Comment (dmibaranov) [inline encryptme_stats/metrics.py]: `and not peers[-1]`

File: encryptme_stats/metrics.py [hunk 5]
```diff
@@ -80,23 +80,51 @@ def _get_openvpn_stats(path="/var/run/openvpn/server-0.sock"):
...
+                    for peer in peers[1:]:
+                        last_handshake = int(peer.split('\t')[4])
+                        # Only count if there was a handshake in the last 20 minutes
+                        if (now_epoch - last_handshake) <= (20*60): 
+                            num_wireguard += 1
+    except Exception as exc:
+        logging.debug("Error getting wireguard connections: %s", exc)
+
+    return num_wireguard
+
```
Comment (dmibaranov) [inline encryptme_stats/metrics.py:109]: two empty lines between functions, not 3

File: encryptme_stats/metrics.py [hunk 6]
```diff
@@ -622,22 +650,100 @@ def _get_ipsec_session_stats():
...
 
+def _get_wireguard_stats():
+    """Gather Wireguard statistics."""
+    try:
+        interfaces = subprocess_out(["wg", "show", "interfaces"])
+        if not isinstance(interfaces, list):
+            return
+        if interfaces and interfaces[-1] == '':
+            interfaces.pop()
+        info = []
```
Comment (dmibaranov) [inline encryptme_stats/metrics.py]: please declare at the top of function

File: encryptme_stats/metrics.py [hunk 7]
```diff
@@ -622,22 +650,100 @@ def _get_ipsec_session_stats():
...
+                        'last_handshake': last_handshake,
+                        'logged_at': now_epoch,
+                        'bytes_up': bytes_up,
+                        'bytes_down': bytes_down,
+                        'protocol': 'wireguard',
+                    }
+                })
+    except Exception as exc:
+        logging.debug("Error gathering wireguard bandwwith: %s", exc)
+        return []
```
Comment (dmibaranov) [inline encryptme_stats/metrics.py]: please remove

File: encryptme_stats/metrics.py [hunk 8]
```diff
@@ -622,22 +650,100 @@ def _get_ipsec_session_stats():
...
+            now_epoch = int(datetime.utcnow().timestamp())
+            for peer in peers:
+                _, _, _, private_ip, last_handshake, bytes_up, bytes_down, _ = peer.split('\t')
+                last_handshake = int(last_handshake)
+                if (now_epoch - 120) > last_handshake:
+                    continue
+                info.append({
+                    'stats_type': 'vpn_session',
+                    'vpn_session': {
+                        'public_id': '',
```
Comment (dmibaranov) [inline encryptme_stats/metrics.py]: peer's pubkey should be here

File: encryptme_stats/metrics.py [hunk 9]
```diff
@@ -622,22 +650,100 @@ def _get_ipsec_session_stats():
...
         openvpn_stat = _get_openvpn_session_stats()
         ipsec_stat = _get_ipsec_session_stats()
+        wireguard_stat = _get_wireguard_stats()
 
-        result = openvpn_stat + ipsec_stat
-        if len(result) == 0:
-            return empty
-
-        return result
+        return (openvpn_stat + ipsec_stat + wireguard_stat)
```
Comment (dmibaranov) [inline encryptme_stats/metrics.py]: please return empty list if there is no stats. that prevents us to spam stats server with empty payloads

File: encryptme_stats/metrics.py [hunk 10]
```diff
@@ -622,22 +650,100 @@ def _get_ipsec_session_stats():
...
     except Exception as exc:
         logging.debug("Error gathering vpn_session stats: %s", exc)
-        return empty
\ No newline at end of file
+        return {}
```
Comment (dmibaranov) [inline encryptme_stats/metrics.py]: should be empty list

File: encryptme_stats/metrics.py [hunk 11]
```diff
@@ -622,22 +650,100 @@ def _get_ipsec_session_stats():
...
+
+
+def wireguard():
+    """Gather Wireguard statistics."""
+    try:
+        interfaces = subprocess_out(["wg", "show", "interfaces"])
+        if not isinstance(interfaces, list):
+            return []
+        if interfaces and interfaces[-1] == '':
```
Comment (dmibaranov) [inline encryptme_stats/metrics.py]: `and not interfaces[-1]:`

File: encryptme_stats/metrics.py [hunk 12]
```diff
@@ -80,23 +80,51 @@ def _get_openvpn_stats(path="/var/run/openvpn/server-0.sock"):
...
+                peers = subprocess_out(["wg", "show", interface, "dump"])
+                if isinstance(peers, list):
+                    if peers and peers[-1] == '':
+                        peers.pop()
+                    if len(peers) < 2:
+                        continue
+                    for peer in peers[1:]:
+                        last_handshake = int(peer.split('\t')[4])
+                        # Only count if there was a handshake in the last 20 minutes
+                        if (now_epoch - last_handshake) <= (20*60):
```
Comment (dmibaranov) [inline encryptme_stats/metrics.py]: `20 * 60`

File: encryptme_stats/metrics.py [hunk 13]
```diff
@@ -622,22 +650,100 @@ def _get_ipsec_session_stats():
         return []
 
 
+def _get_wireguard_stats():
+    """Gather Wireguard statistics."""
+    try:
+        interfaces = subprocess_out(["wg", "show", "interfaces"])
+        if not isinstance(interfaces, list):
+            return
+        if interfaces and interfaces[-1] == '':
```
Comment (dmibaranov) [inline encryptme_stats/metrics.py]: `and not interfaces[-1]:`

**Review by dmibaranov — COMMENTED** [review]
Reviewed

File: encryptme_stats/metrics.py (hunk 1 above)
Comment (fernando-guerrero-leon) [reply encryptme_stats/metrics.py]: Done

File: encryptme_stats/metrics.py (hunk 2 above)
Comment (fernando-guerrero-leon) [reply encryptme_stats/metrics.py]: Done

File: encryptme_stats/metrics.py (hunk 3 above)
Comment (fernando-guerrero-leon) [reply encryptme_stats/metrics.py]: Done

File: encryptme_stats/metrics.py (hunk 4 above)
Comment (fernando-guerrero-leon) [reply encryptme_stats/metrics.py]: Done

File: encryptme_stats/metrics.py (hunk 5 above)
Comment (fernando-guerrero-leon) [reply encryptme_stats/metrics.py:109]: Done

**Comment by fernando-guerrero-leon** [comment]
All requested changes were applied.

**Pull request merged**